Addressing client enumeration section with protocol changes (#156)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

OPAQUE relies on the following protocols and primitives:

- Oblivious Pseudorandom Function (OPRF, {{I-D.irtf-cfrg-voprf}}, version -06):
  - Blind(x): Convert input `x` into an element of the OPRF group, randomize it
    by some scalar `r`, producing `M`, and output (`r`, `M`).
  - Evaluate(k, M): Evaluate input element `M` using private key `k`, yielding
    output element `Z`.
  - Finalize(x, r, Z): Finalize the OPRF evaluation using input `x`,
    random scalar `r`, and evaluation output `Z`, yielding output `y`.
  - SerializedElement: A serialized OPRF group element, a byte array of fixed
    length.
  - SerializedScalar: A serialized OPRF scalar, a byte array of fixed length.
  - Nok: The size of an OPRF private key

> - Deterministic Key Pair Generation Function:

Author: Yes, it is. Can you clarify what you mean by "move it up"? It is already under the OPRF section -- do you mean to move it up in front of `Blind(x)`, `Evaluate(k, M)`, etc?

> - Deterministic Key Pair Generation Function:

Reviewer: Yep! If it's an OPRF API, it should be in the same list as Blind() etc.

> - Deterministic Key Pair Generation Function:

Author: I guess what I am saying is, it is **already** in the same list as Blind() etc... unless I am misinterpreting what you are saying?

> - Deterministic Key Pair Generation Function:

Reviewer: It's in a separate list, like the KDF and other functions below.

> - Deterministic Key Pair Generation Function:

Author: OH I see what you mean -- I thought you were referring to the `Nok` parameter as needing to move up. But now I realize that you are referring to `DeriveKeyPair` being needed to move up. Got it! :) The comment line numbers this points to was a bit ambiguous :)

> - Deterministic Key Pair Generation Function:

Reviewer: Ah, yep :) Sorry, I should have been more clear.

> - Deterministic Key Pair Generation Function:

Author: This should be resolved in the latest diff.

Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.

## Setup Phase {#setup-phase}

In a setup phase, C chooses its password, and S chooses its own pair of private-public
AKE keys (server_private_key, server_public_key) for use with the AKE, along with a Nh-byte oprf_seed. S can use
the same pair of keys with multiple clients, and can opt to use multiple seeds (so long as they are
kept consistent for each client). These steps can happen offline, i.e., before the registration phase.

Once complete, the registration process proceeds as follows.

## Credential Registration

~~~
 Client (password, creds)            Server (server_private_key, server_public_key, oprf_seed)
 --------------------------------------------------------------------
 (request, blind) = CreateRegistrationRequest(password)

                               request
                      ------------------------->

>             (response, oprf_key) = CreateRegistrationResponse(request, server_public_key, credential_identifier, oprf_seed)

Reviewer: Where does `credential_identifier` come from? It should probably be included in the Server inputs on L419.

Author: Added!

Registration is executed between a client C and a
server S. It is assumed S can identify C and the client can
authenticate S during this registration phase. This is the only part
in OPAQUE that requires an authenticated and confidential channel, either physical, out-of-band,
PKI-based, etc. This section describes the registration flow, message encoding,
and helper functions. Moreover, C has a key pair (client_private_key, client_public_key) for an AKE protocol
which is suitable for use with OPAQUE; See {{online-phase}}. The private-public keys (client_private_key, client_public_key) may be randomly generated (using a cryptographically secure pseudorandom number generator) for the account or provided by the calling client.
Clients MUST NOT use the same key pair (client_private_key, client_public_key) for two different accounts.

## Setup Phase {#setup-phase}

In a setup phase, C chooses its password, and S chooses its own pair of private-public
AKE keys (server_private_key, server_public_key) for use with the AKE, along with a Nh-byte oprf_seed. S can use

> the same pair of keys with multiple clients. These steps can happen offline, i.e.,

Reviewer 2: It does not have to be one seed for all. For example, there could be a seed for all userid's that start with letter a, one for those that start with letter b, etc. But a userid should always use the same oprf seed. A change of seed can only happen at re-registration.

> the same pair of keys with multiple clients. These steps can happen offline, i.e.,

Author: @chris-wood: How should we allow for the flexibility in this API, in terms of producing test vectors, etc. ? Does it suffice to just add a note here saying that multiple oprf_seeds can be used?

> the same pair of keys with multiple clients. These steps can happen offline, i.e.,

Reviewer: Yeah, I think noting that multiple seeds may be used is sufficient, but the test vectors should just use one seed for all users.

> the same pair of keys with multiple clients. These steps can happen offline, i.e.,

Author: Added a side note that multiple seeds can be used

~~~
struct {
    opaque client_public_key[Npk];
    opaque masking_key[Nh];
    Envelope envelope;
} RegistrationUpload;
~~~

client_public_key
: The client's encoded public key, corresponding to the private key `client_private_key`.

masking_key

> : A key used by the server to preserve the confidentiality of the envelope during login

Reviewer: ```suggestion
: A key used by the server to preserve confidentiality of the envelope during login
```

### CreateRegistrationResponse {#create-reg-response}

~~~
CreateRegistrationResponse(request, server_public_key, cred_identifier, oprf_seed)

Input:
- request, a RegistrationRequest structure
- server_public_key, the server's public key
- cred_identifier, an identifier that uniquely represents the credential being
  registered
- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key

Output:
- response, a RegistrationResponse structure
- oprf_key, the per-client OPRF key known only to the server

Steps:

> 1. oprf_key = HKDF-Expand(oprf_seed, concat("oprf key", cred_identifier), Nok)

Reviewer: I don't think this is a workable replacement. This basically type casts bytes to an OPRF private key. I think we should revert this, and instead use DeriveKeyPair(HKDF(...)).

Author: Addressed in the latest push

### CreateRegistrationResponse {#create-reg-response}

~~~
CreateRegistrationResponse(request, server_public_key, credential_identifier, oprf_seed)

Input:
- request, a RegistrationRequest structure
- server_public_key, the server's public key
- credential_identifier, an identifier that uniquely represents the credential being
  registered
- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key

Output:
- response, a RegistrationResponse structure
- oprf_key, the per-client OPRF key known only to the server

Steps:

> 1. (oprf_key, _) = DeriveKeyPair(Expand(oprf_seed, concat("oprf key", credential_identifier), Nok))

Author: Thanks, used "OprfKey" here instead, following the usual convention of CamelCase for acronyms that are longer than 2 letters.

Steps:
1. y = Finalize(password, blind, response.data)
2. envelope_nonce = random(32)

> 3. prk = Extract(salt=0, Harden(y, params))

Reviewer: ```suggestion
3. prk = Extract("", Harden(y, params))
```

Author: Cool, also changed the salt=0 instance to "" in the "OPAQUE-3DH Key Schedule" section

Reviewer 2: Actually, it is better to set the salt to 0. It does not have to be described as salt=0 but just 0 (instead of ""). In principle, any fixed salt value can do, but I prefer to see 0 being used as the default "fixed value". In particular, 0 is used in TLS 1.3 key derivation when a random value for salt is not available and here we want to stay as close as possible to TLS 1.3 key derivation.

Reviewer: For what it's worth, "" is the notation used in the TLS zero-length strings, which is what we want here, I think.

Reviewer 2: What is the meaning of 0 in the following TLS 1.3 figure. Is it the value 0 or a 0-length string?
             0
             |
             v
   PSK ->  HKDF-Extract = Early Secret

Reviewer 2: Actually I see the answer: "0" indicates a string of Hash.length bytes set to zero.

We should specify the same here.

Reviewer: That's a zero-length string (or NULL, or the equivalent of providing no salt at all), but I was referring to "" used elsewhere where context strings.

Reviewer: > Actually I see the answer: "0" indicates a string of Hash.length bytes set to zero.

I'm not sure this is best, since the KDF might not be HKDF (and therefore have an implied hash function underneath). I think nil/""/0/whatever are equivalent, and "" seems to be most clear (to me) and consistent with 8446.

Reviewer 2: Good point. If we are treating the KDF as a general Extract-then-Expand primitive then we cannot say "a string of Hash.length bytes set to zero" but we also cannot say the empty string as it may not make sense in some other KDF. If you want to be generic, then you should refer to this salt value as a "constant" (as opposed to a random one), maybe call it const_salt and set it to "a string of Hash.length bytes set to zero" in the case of HKDF.

### Credential File {#credential-file}

> The server then directly stores the `record` object as the credential file for each client. Note that

Author: Folded!

> the values `oprf_seed` and `server_private_key` from the server's setup phase must also be persisted.

Author: I don't think that's necessary. In the same way we don't mention the storing of any user or server identifiers when CustomIdentifier mode is used, I don't think we need to mention that the credential file has to store the credential identifier. But we can add it too if you think it increases clarity

> the values `oprf_seed` and `server_private_key` from the server's setup phase must also be persisted.

Reviewer: I'm mainly interested in reducing the number of parameters to our core functions, which is growing little by little. We can keep this as-is for now, but we should keep this possible change on the back burner.

~~~
struct {
    SerializedElement data;

>     opaque encrypted_response[Npk + Nsk + Nh + 33];

Author: It is. Thanks for the catch -- added!

#### CreateCredentialResponse {#create-credential-response}

There are two scenarios to handle for the construction of a CredentialResponse object: either the
credential_file record for the client exists (corresponding to a properly registered client), or
it was never created (corresponding to a client that has yet to register).

In the case of an existing credential_file record with the corresponding identifier
`credential_identifier`, the server invokes the following function to
produce a CredentialResponse:

~~~
CreateCredentialResponse(request, server_public_key, credential_file, credential_identifier, oprf_seed)

Input:
- request, a CredentialRequest structure
- server_public_key, the public key of the server

> - credential_file, an instance of RegistrationUpload which is the server's

Reviewer: ```suggestion
- record, an instance of RegistrationUpload which is the server's
```

Reviewer: (here and elsewhere, since we refer to the element of this type as `record` in the FinalizeRequest function.)

Author: Done!
